realm/realm-library/src/main/java/io/realm/HandlerController.java

```diff
@@ -146,6 +151,23 @@ public boolean handleMessage(Message message) {
         return true;
     }
 
+    /**
+     * Properly handles when an async transaction completes. This will be treated as a REALM_CHANGED event when
+     * determining which queries to re-run and when to notify listeners.
+     * <p>
+     * NOTE: This is needed as it is not possible to combine a `Message.what` value and a callback runnable. So instead
+     * of posting two messages, we post a runnable that runs this method. This means it is possible to interpret
+     * `REALM_CHANGED + Runnable` as one atomic message.
+     *
+     * @param onSuccess onSuccess callback to run for the async transaction that completed.
+     */
+    public void handleAsyncTransactionCompleted(Runnable onSuccess) {
+        if (onSuccess != null) {
+            pendingOnSuccessAsyncTransactionCallbacks.add(onSuccess);
+        }
+        realmChanged(false);
```

Comment:
Mostly it was to keep the idea that an asyncTransactionCompleted event could function as an atomic event in the message queue which helped avoiding introducing yet another thread safe list, but I agree that because Handler does not support Runnables very well, the code gets a bit ugly. 

The above example you have is kinda likely though and might cause a lot of un-needed change events to happen, so perhaps going back to a CopyOnWriteArrayList is better.

Anyone know roughly where a ReadWriteLock is better than a CopyOnWriteList? Because in this case we would probably write more to it than we would iterate.